sos/collector/transports/juju.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2023 Canonical Ltd., Chi Wai Chan <chiwai.chan@canonical.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+
+import subprocess
+
+from sos.collector.exceptions import JujuNotInstalledException
+from sos.collector.transports import RemoteTransport
+from sos.utilities import sos_get_command_output
+
+
+class JujuSSH(RemoteTransport):
+    """
+    A "transport" that leverages `juju ssh` to perform commands on the remote
+    hosts.
+
+    This transport is expected to be used in juju managed environment, and the
+    user should have the necessary credential for accessing the controller.
+    When using this transport, the --nodes option will be expected to be a
+    comma separated machine IDs, **not** IP addr, since `juju ssh` identifies
+    the ssh target by machine ID.
+
+    Examples:
+
+    sos collect --nodes 0,1,2 --no-local --transport juju --batch
+
+    """
+
+    name = "juju_ssh"
+
+    def _check_juju_installed(self):
+        cmd = "juju version"
+        try:
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
+        except subprocess.CalledProcessError:
+            self.log_error("Failed to check `juju` version")
+            raise JujuNotInstalledException
+        return True
+
+    def _chmod(self, fname):
+        cmd = f"{self.remote_exec} chmod o+r {fname}"
+        try:
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
+        except subprocess.CalledProcessError:
+            self.log_error(f"Failed to make {fname} world-readable")
+            raise
+        return True
+
+    def _connect(self, password=""):
+        self._connected = self._check_juju_installed()
+        return self._connected
+
+    def _disconnect(self):
+        return True
+
+    @property
+    def connected(self):
+        return self._connected
+
+    @property
+    def remote_exec(self):
+        model, target_option = self.address.split(":")
+        model_option = f"-m {model}" if model else ""
+        option = f"{model_option} {target_option}"
+        return f"juju ssh {option} sudo"
```

Comment:
Has this been tested as a non-root connecting user? Our current flows will automatically leverage `sudo` where needed, so I think this could result in a command of `juju ssh sudo sudo <cmd>...`.

Reply:
One part I am not very sure, when will sos leverage `sudo`?
Because we didn't encounter this situation in our testing. Usually the user of juju cli is not root, and the remote juju agent will be root.

Reply:
`sudo` is used if we're connecting as a non-root user and `need_root=True` is passed to `run_command()` for nodes.

On a second review, I think we're skirting the issue here since `ssh_user` is defaulting to `root`, and this transport is just ignoring that option.

I think I'm ok with baking the use of `sudo` directly into command execution, but it's something we'll want to be mindful of once this gets released to end users.

Reply:
Sorry I gave wrong information in my previous comment.

Since the default ssh_user for juju is ubuntu, we need `--ssh-user ubuntu` to successful run collect. I want to provide this option as default so user won't need to provide it every time but I couldn't find a way to add it.

So one simple way is we force add sudo in `remote_exec`, which is fine in juju because the user have nopasswd sudo permission. And we cheat sos the user is root. Which is not a pretty way.

For example:
- if I run with `--ssh-user` `sos collect --cluster-type juju --no-local -c "juju.apps=ubuntu" --ssh-user ubuntu`
- The command run on remote machine is `juju ssh  0 sudo '/bin/bash -c '"'"'sudo -S /usr/bin/sosreport --batch  --chroot auto'"'"''`
- Which is still work to collect.

I think we can discuss is there any more clean way.

Reply:
Hmm - ok, I understand better now.

Right now, we rely solely on the `ssh_user` value to detect if we need to leverage sudo (or, `su -c` if we instead want to directly leverage `root` when `--become` is used). [This is set here in commons](https://github.com/sosreport/sos/blob/main/sos/collector/__init__.py#L535). This is mainly due to the fact that we use the same user across all connected nodes, so running something akin to `whoami` on every node could be seen as superfluous since we allegedly already have the information we need.

A wrapper that automatically assumes its own specific user as well throws that off a bit, but it should be fairly straight forward to handle. We have a couple options, I think:

1. Do a `whoami` or similar check on each node during node `SosNode` initialization. Not a huge fan of this due to the reason above, but it would be easy and serviceable.
2. Within `JujuSSH`, add logic to `remote_exec` to handle the `need_root` parameter for the calling `run_command()`. This is technically possible, but likely very fragile and generally hacky.
3. Accept the default `need_sudo` setting from commons, but allow a node to override this based on the loaded transport. Basically we could add something like `default_user` to transports (default `''` or `None` or what have you) and if it's set, we align that like so:

```python
class SosNode():

def __init__(self, address, commons, password=None, local_sudo=None,
             load_facts=True):
    [...]
    self._transport = self._load_remote_transport(commons)
    if self._transport.default_user:
        self.need_sudo = self._transport.default_user != 'root'
```

I haven't tested the above, but something like that should suffice. I like this approach more because it allows an easy entry point to control this for current and future transports, not just specifically juju.

Reply:
This is works for juju. PR updated.